### PR TITLE
Avoid race condition when updating equivalence cache

### LINKED ISSUE #58507: Need to fix data race when updating equivalence cache
Fix this TODO (raised by @bsalamat) : https://github.com/kubernetes/kubernetes/blob/30b89d830b7fd0827576853d6a0db44b66a90d3d/pkg/scheduler/core/generic_scheduler.go#L482

Part of: #58222  

One thoughts to fix this is, we can redesign the equivalence cache data structure to: 
1. Set `invalid` bool instead of deleting the entry. 
2. Include certain version tag in the entry, update the version per WRITE op (Update and Invalidate). 
2. Return the version after eCache predicate,  and check it before update eCache.

### PR DESCRIPTION
**Pull request by tossmilestone**
**What this PR does / why we need it**:
Lock the ecache to update the ecache on each predicate running, to avoid race condition.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fix #58507 

**Special notes for your reviewer**:
None

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by tossmilestone** [comment]
ping @resouer @bsalamat

**Review by resouer — CHANGES_REQUESTED** [review]
Thanks but you may missed the related issue: https://github.com/kubernetes/kubernetes/issues/58507

But I would be glad to review if you can update this PR to latest design, the change is straightforward.

Please let me know :)

**Comment by tossmilestone** [comment]
OK, I will try to implement the lock design. @resouer

**Comment by tossmilestone** [comment]
@resouer updated, PTAL:)

**Comment by resouer** [comment]
Will take a look tomorrow.

File: pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -488,8 +491,15 @@ func podFitsOnNode(
 						} else {
 							predicateResults[predicateKey] = HostPredicate{Fit: fit, FailReasons: reasons}
 						}
+						result := predicateResults[predicateKey]
+						ecache.UpdateCachedPredicateItem(pod.GetName(), info.Node().GetName(), predicateKey, result.Fit, result.FailReasons, equivCacheInfo.hash, false)
 					}
 				}
+
+				if eCacheAvailable {
+					ecache.Unlock()
```
Comment (resouer) [inline pkg/scheduler/core/generic_scheduler.go]: Why not release lock just following `UpdateCachedPredicateItem()`? 

Also, I am not sure about purpose of `needLock bool`, seems like you can just move lock control out from ecache.

You may want to still use defer to ensure  `ecache.Unlock()`.

Comment (tossmilestone) [reply pkg/scheduler/core/generic_scheduler.go]: >Emm ... why not release lock just following UpdateCachedPredicateItem() above?

Lock from `ecache.PredicateWithECache` to `UpdateCachedPredicateItem` to ensure that the result of `PredicateWithECache` is not changed when update the ecache.

>Also, I am not sure about purpose of needLock bool, seems like you can just move lock control out from ecache.

I just suppose there may be requirements in future for using ` UpdateCachedPredicateItem()` with lock inside the method.

@resouer

**Comment by bsalamat** [comment]
Looks good to me, but I would like @resouer  to take a final look.

**Review by resouer — APPROVED** [review]
The only nit I saw is `needLock bool` which seems unnecessary, but as mentioned it's for future change consideration, I am OK to have it.

**Comment by tossmilestone** [comment]
@bsalamat everything is OK now, just needs your approve. Thanks!

**Comment by bsalamat** [comment]
/lgtm

Thanks, @tossmilestone!

**Pull request merged**